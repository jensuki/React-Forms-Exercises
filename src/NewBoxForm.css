.form-group {
    display: flex;
    align-items: center;
    justify-content: center;
    margin-bottom: 10px;
}

label {
    width: 120px;
    text-align: right;
    margin-right: 10px;
}

button {

    margin: 20px auto;
}