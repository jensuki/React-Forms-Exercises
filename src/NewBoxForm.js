import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import './NewBoxForm.css'

const NewBoxForm = ({ addBox }) => {
    const INITIAL_STATE = {
        width: '',
        height: '',
        backgroundColor: ''
    }

    const [formData, setFormData] = useState(INITIAL_STATE);
    const [isEmpty, setIsEmpty] = useState(false);

    // handle form input changes
    const handleChange = (e) => {
        const { name, value } = e.target;
        setFormData(data => ({
            ...data,
            [name]: value
        }))
    }

    // handle form submission
    const handleSubmit = (e) => {
        e.preventDefault();

        // check if any fields are empty
        if (!formData.width || !formData.height || !formData.backgroundColor) {
            setIsEmpty(true);
            return;
        }

        addBox({ ...formData, id: uuid() }) // add new box with unique id
        setFormData(INITIAL_STATE); // clear form after submit
    }
    return (
        <form onSubmit={handleSubmit}>
            <div className="form-group">
                <label htmlFor="width">Width:</label>
                <input
                    type="text"
                    name="width"
                    value={formData.width}
                    onChange={handleChange}
                    id="width"
                />
            </div>
            <div className="form-group">
                <label htmlFor="height">Height:</label>
                <input
                    type="text"
                    name="height"
                    value={formData.height}
                    onChange={handleChange}
                    id="height"
                />
            </div>
            <div className="form-group">
                <label htmlFor="backgroundColor">Background Color:</label>
                <input
                    type="text"
                    name="backgroundColor"
                    value={formData.backgroundColor}
                    onChange={handleChange}
                    id="backgroundColor"
                />
            </div>
            {isEmpty && <p style={{ color: 'red' }}>All fields are required</p>}
            <button>Add Box</button>


        </form>
    )
}

export default NewBoxForm;