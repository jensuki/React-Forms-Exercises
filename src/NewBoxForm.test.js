import { render, fireEvent, screen } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

it('renders without crashing', () => {
    render(<NewBoxForm addBox={() => { }} />)
})

it('matches snapshot', () => {
    const { asFragment } = render(<NewBoxForm addBox={() => { }} />)
    expect(asFragment()).toMatchSnapshot();
})

it('updates form fields on user input', () => {
    render(<NewBoxForm addBox={() => { }} />);

    const widthInput = screen.getByLabelText('Width:');
    const heightInput = screen.getByLabelText('Height:');
    const backgroundColorInput = screen.getByLabelText('Background Color:');

    fireEvent.change(widthInput, { target: { value: '200' } });
    fireEvent.change(heightInput, { target: { value: '100' } });
    fireEvent.change(backgroundColorInput, { target: { value: 'blue' } });

    expect(widthInput.value).toBe('200');
    expect(heightInput.value).toBe('100');
    expect(backgroundColorInput.value).toBe('blue');

})