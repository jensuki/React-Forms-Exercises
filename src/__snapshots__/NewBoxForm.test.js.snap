// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="form-group"
    >
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
    </div>
    <div
      class="form-group"
    >
      <label
        for="backgroundColor"
      >
        Background Color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        type="text"
        value=""
      />
    </div>
    <button>
      Add Box
    </button>
  </form>
</DocumentFragment>
`;
